src/attach-shadow.js

```diff
@@ -237,25 +293,141 @@ ShadyRoot.prototype._updateChildNodes = function(container, children) {
     for (let j=s.index, n; j < s.index + s.addedCount; j++) {
       n = children[j];
       insertBefore.call(container, n, next);
-      // TODO(sorvell): is this splice strictly needed?
       composed.splice(j, 0, n);
     }
   }
 }
 
-ShadyRoot.prototype.getInsertionPointTag = function() {
-  return this._distributor.insertionPointTag;
+/**
+ * Adds the given slots. Slots are maintained in an dom-ordered list.
+ * In addition a map of name to slot is updated.
+ */
+ShadyRoot.prototype._addSlots = function(slots) {
+  let slotNamesToSort;
+  for (let i=0; i < slots.length; i++) {
+    let slot = slots[i];
+    // ensure insertionPoints's and their parents have logical dom info.
+    // save logical tree info
+    // a. for shadyRoot
+    // b. for insertion points (fallback)
+    // c. for parents of insertion points
+    slot.__shady = slot.__shady || {};
+    recordChildNodes(slot);
+    recordChildNodes(slot.parentNode);
+    let name = this._nameForSlot(slot);
+    if (this._slotMap && this._slotMap[name]) {
+      slotNamesToSort = slotNamesToSort || {};
+      slotNamesToSort[name] = true;
+    }
+    this._slotList.push(slot);
+  }
+  if (slotNamesToSort) {
+    for (let n in slotNamesToSort) {
+      let slotsForName = this._extractSlotsOfName(n);
+      let sorted = this._sortSlots(slotsForName);
+      this._slotList.push(...sorted);
```

Comment:
Would map of sorted list of slots be a better data structure to use than list + map?
e.g.
```js
_slotMap = {
  'header': [<slot name="header">, <slot name="header">],
  'footer': [<slot name="footer">, <slot name="footer">]
}
```

* When adding a slot you just do `_slotMap[name].push(slot)`
* If `_slotMap[name].length > 0` before adding, also add to `slotNamesToSort`
* When there are `slotNamesToSort`, it's just `_slotMap[name] = sorted`
* When slotting you use `_slotMap[name][0]`.

Would help eliminate the `_extractSlotsOfName` because you can just sort the array for that slot name.  OTOH I guess there are a handful of spots where all slots need to be iterated, which would make it two loops instead of one.  So not slam dunk.